FFmpegInterop/SampleProvider.cpp

```diff
@@ -0,0 +1,415 @@
+
+#include "pch.h"
+#include "SampleProvider.h"
+#include "Reader.h"
+
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Storage::Streams;
+using namespace winrt::Windows::Media::Core;
+using namespace winrt::Windows::Media::MediaProperties;
+using namespace std;
+
+namespace winrt::FFmpegInterop::implementation
+{
+	SampleProvider::SampleProvider(_In_ const AVStream* stream, _In_ Reader& reader) :
+		m_stream(stream),
+		m_reader(reader)
+	{
+		if (m_stream->start_time != AV_NOPTS_VALUE)
+		{
+			m_startOffset = m_stream->start_time;
+			m_nextSamplePts = m_stream->start_time;
+		}
+	}
+
+	void SampleProvider::SetEncodingProperties(_Inout_ const IMediaEncodingProperties& encProp, _In_ bool setFormatUserData)
+	{
+		const AVCodecParameters* codecPar{ m_stream->codecpar };
+
+		switch (codecPar->codec_type)
+		{
+		case AVMEDIA_TYPE_AUDIO:
+		{
+			IAudioEncodingProperties audioEncProp{ encProp.as<IAudioEncodingProperties>() };
+			audioEncProp.SampleRate(codecPar->sample_rate);
+			audioEncProp.ChannelCount(codecPar->channels);
+			audioEncProp.Bitrate(static_cast<uint32_t>(codecPar->bit_rate));
+			audioEncProp.BitsPerSample(static_cast<uint32_t>(max(codecPar->bits_per_coded_sample, codecPar->bits_per_raw_sample)));
+
+			if (setFormatUserData && codecPar->extradata != nullptr && codecPar->extradata_size > 0)
+			{
+				// Set the format user data
+				IAudioEncodingPropertiesWithFormatUserData audioEncPropWithFormatUserData{ encProp.as<IAudioEncodingPropertiesWithFormatUserData>() };
+				audioEncPropWithFormatUserData.SetFormatUserData({ codecPar->extradata, codecPar->extradata + codecPar->extradata_size });
+			}
+
+			MediaPropertySet properties{ audioEncProp.Properties() };
+
+			if (GUID subtype{ unbox_value<GUID>(properties.Lookup(MF_MT_SUBTYPE)) }; subtype == MFAudioFormat_PCM || subtype == MFAudioFormat_Float)
+			{
+				properties.Insert(MF_MT_ALL_SAMPLES_INDEPENDENT, PropertyValue::CreateUInt32(true));
+				properties.Insert(MF_MT_COMPRESSED, PropertyValue::CreateUInt32(false));
+			}
+			else
+			{
+				properties.Insert(MF_MT_ALL_SAMPLES_INDEPENDENT, PropertyValue::CreateUInt32(false));
+				properties.Insert(MF_MT_COMPRESSED, PropertyValue::CreateUInt32(true));
+			}
+			
+			break;
+		}
+
+		case AVMEDIA_TYPE_VIDEO:
+		{
+			IVideoEncodingProperties videoEncProp{ encProp.as<IVideoEncodingProperties>() };
+			videoEncProp.Height(codecPar->height);
+			videoEncProp.Width(codecPar->width);
+			videoEncProp.Bitrate(static_cast<uint32_t>(codecPar->bit_rate));
+
+			if (codecPar->sample_aspect_ratio.num > 0 && codecPar->sample_aspect_ratio.den != 0)
+			{
+				MediaRatio pixelAspectRatio{ videoEncProp.PixelAspectRatio() };
+				pixelAspectRatio.Numerator(codecPar->sample_aspect_ratio.num);
+				pixelAspectRatio.Denominator(codecPar->sample_aspect_ratio.den);
+			}
+
+			if (m_stream->avg_frame_rate.num != 0 || m_stream->avg_frame_rate.den != 0)

```

Comment:
>|| [](start = 41, length = 2)

Should this be &&?

Reply:
That's what was there before, but yeah I think you're right.